hostapd: add sae_password_file support

roshii opened the pull request:
With `sae` or `sae-mixed` encryption it is currently not possible to assign multiple passwords to a single `bss`.

This PR adds multiple passwords support for the above mentioned encryption mode both via `wifi-station` config section and `sae_password_file` entry on `wifi-iface` section, therefore allowing VLAN ID assignment for WPA3 clients (using password identifiers).

rany2 commented on package/network/services/hostapd/files/hostapd.sh:
```diff
@@ -408,7 +409,7 @@ hostapd_set_psk_file() {
 	set_default mac "00:00:00:00:00:00"
 	[ -n "$vid" ] && vlan_id="vlanid=$vid "
-	echo "${vlan_id} ${mac} ${key}" >> /var/run/hostapd-${ifname}.psk
+	echo "${vlan_id}${mac} ${key}" >> /var/run/hostapd-${ifname}.psk
```
This is wrong, there needs to be a space between vlanid and macaddr. This won't work.

roshii replied:
A space character is already set at the previous line: if `vlanid` is set `vlan_id` will be inserted followed by a space, if not a space will not be inserted before `mac`.

rany2 replied:
I missed that, you're right. Maybe it's personal preference but I think it's nicer to remove the space from $vlan_id and keep the space in the echo command. I find it clearer that way.

roshii replied:
That was my idea first, I find it clearer as well.
While not a problem per se it will insert a leading space before `mac` when `vid` is empty... Alternatively, it could be modified as follows:

```bash
set_default mac "00:00:00:00:00:00"
local psk_file_entry="${mac} ${key}"
[ -n "$vid" ] && psk_file_entry="vlanid=${vid} ${psk_file_entry}"
echo "${psk_file_entry}" >> /var/run/hostapd-${ifname}.psk
```

rany2 replied:
I'd just remove this change to keep the scope towards sae_password_file, though I personally think it would be much better to just backport the patch that added sae_password_file in hostapd instead of this.

rany2 replied:
P.S. if you don't know how, you could cherry-pick this commit: https://github.com/rany2/openwrt-1/commit/3788bf8e602c264db18737de9e65521d37bc0aed

yogo1212 replied:
Just throwing it in:
```
hostapd_set_psk_file() {
        ..
        [ -n "$vid" ] && echo -n "vlanid=${vid} "
        echo "${mac} ${key}"
}

hostapd_set_psk() {
	local ifname="$1"
	
	for_each_station hostapd_set_psk_file "${ifname}" > /var/run/hostapd-${ifname}.psk
}
```

Free for the taking :wink:

rany2 commented on package/network/services/hostapd/files/hostapd.sh:
```diff
@@ -418,6 +419,25 @@ hostapd_set_psk() {
+	local sae_password="${key}"
+	[ -n "$mac" ] && sae_password="${sae_password}|${mac}"
+	[ -n "$vid" ] && sae_password="${sae_password}|vlanid=${vid}"
+	echo "${sae_password}" >> /var/run/hostapd-${ifname}.sae_passwords
```
It's a nitpick but because the key will be at the start of the line and it could be anything (including a `-`), I think a `printf '%s\n' "${sae_password}"` is more appropriate.

rany2 commented on package/network/services/hostapd/files/hostapd.uc:
```diff
@@ -280,10 +281,14 @@ function find_array_idx(arr, key, val)
 
 function bss_reload_psk(bss, config, old_config)
 {
-	if (is_equal(old_config.hash.wpa_psk_file, config.hash.wpa_psk_file))
```
This shouldn't be here, it should instead regenerate the config if the sae_password_file is different. This won't do anything.

roshii replied:
I am really not sure about this one but simply put: if it doesn't do anything here, why would it do something in case of change to `wpa_psk_file` ?  Shouldn't the logic be the same for both `sae_password_file` and `wpa_psk_file` ?

rany2 replied:
Because `RELOAD_WPA_PSK` only reloads the wpa_psk_file and it's functionality handled by hostapd itself, `sae_password_file` is not handled by hostapd (as we're generating the sae_password line on the fly) and so we will need to regenerate the config and reload if it changed.

roshii replied:
Commit https://github.com/openwrt/openwrt/pull/14448/commits/b78f011fc09da040a03f038ff4a0b1fb811a51fa is my best guess at solving this one. I never did touch UnrealScript before and so I cannot guarantee syntax is correct but the logic should work

rany2 replied:
It's not UnrealScript, it's ucode: https://github.com/jow-/ucode

roshii replied:
> It's not UnrealScript, it's ucode: https://github.com/jow-/ucode

As you can imagine I did never used ucode either, thanks :)

yogo1212 commented on package/network/services/hostapd/files/hostapd.sh:
```diff
@@ -408,7 +409,7 @@ hostapd_set_psk_file() {
 	json_get_vars mac vid key
 	set_default mac "00:00:00:00:00:00"
```
With SAE, there can only be one default key because - unlike with PSK2 - the AP can't check a PMK on its own.
That means there can be only one definition for MAC `00:00:00:00:00:00`.
Maybe it makes sense to make that more explicit?

yogo1212 replied:
(that is unless we start using `id`?)
hmm

roshii replied:
Password ID support implemented via https://github.com/openwrt/openwrt/pull/14448/commits/e05e0c35fd7c67c576a896c37476624a5e49f916
Exiting function with a message to stdout if a password is already set for default MAC with no ID.

rany2 commented:
Personally I think the approach in my PR is slightly better, https://github.com/openwrt/openwrt/pull/14515. I simply used the upstream patch

yogo1212 commented:
@rany2 If there wasn't already the infrastructure for defining stations using UCI, I'd also favour your more direct approach.
Presumedly, there will be some of both. The `sae_password_file` and support for UCI-defined stations.

We should figure out how UCI and `wpa_psk_file` should work in @Ansuel 's [PR](https://github.com/openwrt/openwrt/pull/14789). Likely, what applies to `wpa_psk_file` will apply to `sae_password_file` as well.

Just adding the option like you do here might make future work more difficult if people start using the file directly (going past the existing Openwrt infrastructure).

rany2 commented:
> Likely, what applies to `wpa_psk_file` will apply to `sae_password_file` as well.

No, it does not. WPS is only a WPA2 thing.

yogo1212 commented:
@rany2 I agree that WPS has nothing to do with it.

The more important point is:
There are two principal ways to specify passwords. One being UCI and a file being another.
How they function together - if they do at all - should be figured out before we put in-place a solution that people start using the file because it might narrow down the possibilities in the future.

For the UCI-based mechanism, the file is just an artifact. It seems to me that your PR intends to have people writing to the file directly. If they start doing that, it's going to be more difficult to start using the UCI-defined stations for SAE in the future.

rany2 commented:
> For the UCI-based mechanism, the file is just an artifact. It seems to me that your PR intends to have people writing to the file directly. If they start doing that, it's going to be more difficult to start using the UCI-defined stations for SAE in the future.

It doesn't matter, if someone wanted to use a UCI-based mechanism they could just specify `sae_password` manually and we would simply create a new UCI option called like `list sae_password 'blah blah blah'`. sae_password_file works in the same way as specifying `sae_password` manually one line at a time, there is no need for it to be an artifact. I think your concern is misplaced.

yogo1212 commented:
I would like to disagree respectfully.
Adding options to specify the list file directly is something that can't be taken back easily.
If it conflicts with a solution that is preferable, it is going to be difficult to take back because there'll be users who's setups are going to break if it's changed.

Also, I would strongly advise against a `list sae_password` option.
With SAE, both parties (AP and station) have to cooperate to make **one** attempt at testing a password.
If you want to test 10 passwords and the last one is correct, the end user will receive **nine** messages telling them the password they entered was wrong despite you expecting them to try the same password again.
And that's just in theory. In practice, hostapd doesn't support that.

There's only one attempt. No guesswork and no brute-force (like it was possible with PSK2). A list of passwords simply makes no sense. You need to know which 1 password to test. In effect, this means you get one dictionary lookup (MAC + opt. key id) and, in case of a miss, you could go with a default.

rany2 commented:
@yogo1212 I really don't understand what you're arguing about. Hostapd already supports specifying sae_password multiple times. You could specify both sae_password and sae_password_file, it's not mutually exclusive. sae_password_file is just treated by hostapd as if each line is a new sae_password entry.

The only difference is that you could only have one wildcard entry with SAE (if you specify multiple wildcard entries, the last one takes effect) whereas PSK does a brute force matching for every matching entry.

yogo1212 commented:
@rany2 In my last comment, I was mainly concerned that there was a misconception about SAE because your proposal of adding a `list` option to specify SAE passwords.
But it appears that is no issue.

My more general concern is that there is currently dichotomy regarding `wpa_psk_file` that is replicated for `sae_password_file` in this PR.

1. it is used in `hostapd.sh` and `mac80211.sh` to load passwords/VLAN IDs from UCI
  - `/var/run/hostapd-${ifname}.psk` 'belongs' to Openwrt
2. users can manually specify `wpa_psk_file` and use their own
  - `wpa_psk_file` belongs to the user
  - the stations configured in UCI become ineffective
3. hostapd writes to `wpa_psk_file` if WPS is enabled
  - `wpa_psk_file` 'belongs' to hostapd
  - doesn't apply directly to `sae_password_file` but the solution for `wpa_psk_file` might work well here too

It makes sense for Openwrt to provide an abstraction for PSK2/SAE and, in fact, it already does (in part).
For that reason, a solution should be expected for `wpa_psk_file`. Maybe that means introducing a breaking change in Openwrt 24.

I simply want to avoid having to do a breaking change here as well. Right now, there is no support for `sae_password_file`, which means there is no breakage - there is also no pressure to introduce a feature prone to 'breaking' in the future.
If a feature is out there, we have to assume there are people using it which likely makes "add a UCI config flag to deeming the usage model for sae_password_file" a breaking change.

One way to deal with this, could be to only use `sae_password_file` internally for UCI-based provisions rather than exposing it to users.
That way, support for user-supplied files can be enabled later according to the model that (hopefully) will be agreed upon.

Another way could be to agree now that the goal will be to always merge UCI  stations+ user-supplied file into one.

It is not my intention to upset anyone now but to avoid upset in the future.

yogo1212 commented on package/network/services/hostapd/files/hostapd.sh:
```diff
@@ -683,15 +704,39 @@ hostapd_set_bss_options() {
+					wireless_setup_vif_failed INVALID_WPA_PSK
+					return 1
+				fi
+				[ -z "$sae_password_file" ] && set_default sae_password_file /var/run/hostapd-$ifname.sae_passwords
```
same here (superfluous `[ -z .. ]`)

rany2 commented on package/network/services/hostapd/files/hostapd.uc:
```diff
@@ -279,23 +279,29 @@ function find_array_idx(arr, key, val)
-	)
+	if (is_equal(type, "sae")) {
+		let file = "sae_password_file"
+		let cmd = "RELOAD_CONFIG"
```
This doesn't work, it doesn't reread the sae_password_file.

roshii replied:
@rany2 if it doesn't reread the file; how can that work for psk?
`bss_reload_psk` is comparing file hashes which must be computed by reading the file and so if it works for `wpa_psk_file` there is no reason it does not for `sae_password_file` or am I missing something?

roshii replied:
if i'm not mistaking hashes are calculated only once, when loading config in `iface_load_config` which is called prior to `iface_set_config` and in which `iface_reload_config` is called and comparing file hashes. so afaik, file hash is updated at the time `bss_reload_psk` is called

rany2 replied:
@roshii I think it's just hostapd bug

rany2 replied:
Oh, you're reading the sae_password_file yourself and then setting the sae_password(s). I thought you were using the hostapd option by now as it is now available in OpenWRT's hostapd. I'm not sure if the bug is relevant if you're setting the sae_password in the hostapd config but maybe not. You need to test it.

However I know for sure that if you used hostapd `sae_password_file` option it does not work

jow- replied:
This code can't work since `let xxx` variables declared inside a block are scoped to that block, they`re not visible outside. Also semicolons are mandatory to terminate a statement (they're only optional for the last statement in a block).

It needs to be something like this:

```javascript
let file, cmd, msg;

if (is_equal(type, "sae")) {
	file = "sae_password_file";
	cmd = "RELOAD_CONFIG";
	msg = "Reload config on sae_password_file change for bss";
}
else {
	file = "wpa_psk_file";
	cmd = "RELOAD_WPA_PSK";
	msg = "Reload WPA PSK file for bss";
}
```

rany2 replied:
Shouldn't ucode complain about this not being valid? At any rate, I was actually referring to RELOAD_CONFIG and RELOAD_BSS_CONFIG not reloading for hostapd's sae_password_file option in specific; I hadn't confirmed that this issue is relevant for when sae_password is defined in hostapd config. I didn't know that this wasn't valid code in the first place. I didn't test this submission.

roshii replied:
> Oh, you're reading the sae_password_file yourself and then setting the sae_password(s). I thought you were using the hostapd option by now as it is now available in OpenWRT's hostapd. I'm not sure if the bug is relevant if you're setting the sae_password in the hostapd config but maybe not. You need to test it.

Thanks, I will test implementation and report back.

roshii replied:
@rany2 I can now confirm `wifi` script needs to be called when sae_password_file is amended. hostapd.conf is then regenerated with sae_password_file content.

roshii commented:
I realize the PR is far from perfect, I am discovering OpenWrt inner working with it. 

> One way to deal with this could be to only use `sae_password_file` internally for UCI-based provisions rather than exposing it to users. That way, support for user-supplied files can be enabled later according to the model that (hopefully) will be agreed upon.
> 
> Another way could be to agree now that the goal will be to always merge UCI stations and user-supplied file into one. Even tacking that on later can be a breaking change if users have come to rely on the fact that UCI-based stations are separate from their self-maintained file.

Both options make sense IMHO. We could proceed as follows
1. Amend this PR to just parse UCI configuration into `sae_password` entries, or (ideally) using `sae_password_file` option with hostapd patch from https://github.com/openwrt/openwrt/pull/14515 which would allow users to use WPA3 just like WAP2
2. Create a new PR that handles merging UCI stations and user-supplied file into one and exposing `sae_password_file` to UCI

It could also be handy to add an `exlude-iface` list option to `wifi-station` and `wifi-vlan` sections - or convert `iface` to a list - to allow users to specify precisely to which interface the config does apply to.

roshii commented:
Latest commit adds support for password identifiers in `wifi-station` section and gives priority to UCI configured password over the ones supplied by file. All password will end-up merged into a single OpenWrt owned file. Would anyone disagree with this approach, I can simply remove the `sae_password_file` option until we come to an agreement on how to handle merging and just keep the UCI defined station.

roshii commented:
I just squashed all commits and will make sure I do not pollute this PR with work-in-progress commits in the future - I clearly made it a mess.

I am really not sure about what to do with the changes to [hostapd.uc](https://github.com/openwrt/openwrt/commit/d636eccb37b7fe87634d090c12d1ed5628001ec2#diff-5244f9939d89fea39a79e4fec575cf9051b6608e25297c81ff67c51cb79d63b0), I guess it should just be removed and aligned with what @rany2 did in https://github.com/openwrt/openwrt/pull/14515. Please advice.

Any suggestions are welcomed.

roshii committed:
hostapd: add support for multiple SAE password

- Adds SAE support to station defined using `wifi-section` configuration section
- Adds support for password identifiers using `pwid` option in `wifi-section` section
- Adds support for `sae_password_file` option in `wifi-iface` section, merging with UCI defined password

Signed-off-by: Simon Castano <roshii@riseup.net>
